core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala

```diff
@@ -41,8 +43,19 @@ class BrokerMetadataListener(
   val maxBytesBetweenSnapshots: Long,
   val snapshotter: Option[MetadataSnapshotter],
   brokerMetrics: BrokerServerMetrics,
-  metadataLoadingFaultHandler: FaultHandler
+  _metadataLoadingFaultHandler: FaultHandler
```

Comment:
The errors handled by this FaultHandler are non-fatal. AFAIK only the active controller will experience fatal metadata errors. For BrokerMetadataListener, we just log an ERROR and increment the metrics.

We call this fault handler in three cases in BrokerMetadataListener:
* Replaying a record within a committed batch
* Replaying a record within a batch from a snapshot
* Finalizing a snapshot

Reply:
Failures in this code seems to be fatal given the latest code:
https://github.com/apache/kafka/blob/cc185807591e54c7abb9af91fbb847ab9a9acc8a/core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala#L323-L330

Reply:
Ok, I see what you mean. It does stand to reason that an error when handling a record would lead to an error when building the MetadataImage itself. Seems like we could cover the `delta.apply` errors with the same non-fatal fault handler.

Reply:
yes, the apply errors should be handled by `metadataLoadingFaultHandler` as well.